### PR TITLE
[metrics_transform_processor] Add filtering capabilities matching metric label values for applying changes

### PR DESCRIPTION
**Pull request by hossain-rayhan**
**Description:** 
This change adds a new feature which will support to filter metrics and apply changes matching metric names and label values together. Earlier, we were able to filter metrics using only metric names and apply our transforms to all the metrics. This enhanced filtering feature will give a new option to apply changes to a certain set of metrics where we have a match for metric name and specific set of metric labels. 

**Can you share a use case?**
Yes. Using prometheus receiver in Kubernetes, we get same metrics (same name) for containers as well as pods. We can differentiate the metrics using metric label values. In our use cases we need to rename some of the pod level metrics without affecting the container level metrics. Hence, we need an option for applying changes to metircs based on metric names and metric label values together.

**Why did you put this change in metricstransformprocessor?**
Metrics transform processor already have the insert and update functions. It applies changes to metrics matching the metric name. Its super easy and meaningful to add couple of lines to the filtering options to match metrics using metric label values besides metric names.

**Testing:** 
Wrote Unit tests and tested manually on local machine.

**Documentation:** 
Updated README with proper config examples.

### CODE REVIEW

**Comment by bogdandrutu** [comment]
Thanks for this PR @hossain-rayhan, unfortunately we don't want to add new functionality to the metrics transform processor until we change it to use the new pdata instead of the old opencensus

**Comment by hossain-rayhan** [comment]
> Thanks for this PR @hossain-rayhan, unfortunately we don't want to add new functionality to the metrics transform processor until we change it to use the new pdata instead of the old opencensus

Hi @bogdandrutu, this is a small addition. I understand the concern but this change truely belongs to this `metricstransformprocessor`. I can give you my word to contribute for changing this part when rewriting the processor. Maybe I can create an issue and assign it to me for this. But for now, this is the fasted path forward for us.

**Comment by hossain-rayhan** [comment]
Hi @bogdandrutu, expecting a second thought from you considering my reply on your comment. Let me know if it sound convincing.

**Comment by hossain-rayhan** [comment]
According to our todays (04/28) Collector SIG meetings decision, we will move forward with this PR. I created an issue to rewrite this part to use OTLP metrics instead of OpenCensus metric type. I will work on this when we plan to rewrite the `metricstransformprocessor`. Let's review this.

Issue Link to track: #3269 
@alolita can you please assign the issue to me? 

Thanks @bogdandrutu and @alolita

**Comment by tigrannajaryan** [comment]
@bogdandrutu re-assigning to you since you discussed this in SIG meeting.

**Comment by hossain-rayhan** [comment]
Hi @bogdandrutu its blocking our release. A review would be highly appreciated.

File: processor/metricstransformprocessor/metrics_transform_processor.go [hunk 1]
```diff
@@ -406,3 +412,45 @@ func (mtp *metricsTransformProcessor) compareTimestamps(t1 *timestamppb.Timestam
...
+					if timeseries.LabelValues[idx].Value != value {
+						return false
+					}
+				}
+				break
+			}
+		}
+
+		// if a label-key is not found and the label-value is non-empty, return false
+		if !keyFound && !(value == "" || isEmptyExp(value)) {
```
Comment (Aneurysm9) [inline processor/metricstransformprocessor/metrics_transform_processor.go:446]: I don't think this logic matches the description in the comment.  It looks like a non-empty label value that is matched by a regexp that could also match an empty label value will hit this condition.  What is intended to be tested here?

Comment (hossain-rayhan) [reply processor/metricstransformprocessor/metrics_transform_processor.go:446]: We have two different use case here. Say, in our config file we have the following labels to match,
1. `{"container": "my-container"}` where the value of the key `container` is non-empty.
2. `{"container": ""}` where the value of the key `container` is empty-string. 

For first case, if we don't find a key, we return false. But for the second case, if we don't find the key, we need to return true which is expected. 

As we don't have a way to make sure a given key is not present, I wrote this logic as alternative. If we want to confirm that a given label is not present, we set the value as empty-string. 

Hope it helps to explain.

Comment (hossain-rayhan) [reply processor/metricstransformprocessor/metrics_transform_processor.go]: Updated the comment to make it more clear.

File: processor/metricstransformprocessor/metrics_transform_processor.go [hunk 2]
```diff
@@ -406,3 +412,45 @@ func (mtp *metricsTransformProcessor) compareTimestamps(t1 *timestamppb.Timestam
...
+
+		for idx, label := range metric.MetricDescriptor.LabelKeys {
+			if label.Key != key {
+				continue
+			}
+
+			keyFound = true
+			for _, timeseries := range metric.Timeseries {
+				if isRegexp {
+					re := regexp.MustCompile(value)
```
[the remark this answers is not in the document]
Comment (hossain-rayhan) [reply processor/metricstransformprocessor/metrics_transform_processor.go]: Updated.

File: processor/metricstransformprocessor/metrics_transform_processor.go [hunk 3]
```diff
@@ -88,15 +91,18 @@ func (f internalFilterStrict) getSubexpNames() []string {
 }
 
 type internalFilterRegexp struct {
-	include *regexp.Regexp
+	include     *regexp.Regexp
+	matchLabels map[string]string
```
[the remark this answers is not in the document]
Comment (hossain-rayhan) [reply processor/metricstransformprocessor/metrics_transform_processor.go]: Updated.

File: processor/metricstransformprocessor/metrics_transform_processor.go [hunk 4]
```diff
@@ -406,3 +412,45 @@ func (mtp *metricsTransformProcessor) compareTimestamps(t1 *timestamppb.Timestam
...
+		return true
+	}
+
+	for key, value := range filterLabels {
+		keyFound := false
+
+		for idx, label := range metric.MetricDescriptor.LabelKeys {
+			if label.Key != key {
+				continue
+			}
```
[the remark this answers is not in the document]
Comment (hossain-rayhan) [reply processor/metricstransformprocessor/metrics_transform_processor.go]: Technically it won't repeat. But as its not a map, we cannot garuntee I think.

**Review by Aneurysm9 — APPROVED** [review]

**Review by mxiamxia — APPROVED** [review]
LGTM!

**Comment by hossain-rayhan** [comment]
@alolita can you please add the `ready_to_get_merged` tag on this. Got approval from Anthony and Min. Thanks.

File: processor/metricstransformprocessor/metrics_transform_processor.go [hunk 5]
```diff
@@ -87,16 +91,50 @@ func (f internalFilterStrict) getSubexpNames() []string {
...
+
+			keyFound = true
+			for _, timeseries := range metric.Timeseries {
+				if timeseries.LabelValues[idx].Value != value {
+					return false
+				}
+				break
```
Comment (anuraaga) [inline processor/metricstransformprocessor/metrics_transform_processor.go:112]: Because of this break, isn't only the first time series checked?

Comment (hossain-rayhan) [reply processor/metricstransformprocessor/metrics_transform_processor.go:112]: Yes. That's I checked for. If the first timeseries matches the label, we will add the metric.

Comment (anuraaga) [reply processor/metricstransformprocessor/metrics_transform_processor.go:112]: Better to be explicit about that than use a loop then.

Comment (hossain-rayhan) [reply processor/metricstransformprocessor/metrics_transform_processor.go:112]: Can you please rephrase for me? I don't think I understand what you meant.

Comment (anuraaga) [reply processor/metricstransformprocessor/metrics_transform_processor.go:112]: If you only want to check the first one can't you use `metric.Timeseries[0]`?

Comment (hossain-rayhan) [reply processor/metricstransformprocessor/metrics_transform_processor.go:112]: Oh I see. Makes sense. Will update.

Comment (hossain-rayhan) [reply processor/metricstransformprocessor/metrics_transform_processor.go]: Updated, thanks.

**Comment by hossain-rayhan** [comment]
Hi @anuraaga I need your approval. Would you please have another look. Thanks.

**Review by anuraaga — APPROVED** [review]

**Comment by hossain-rayhan** [comment]
@tigrannajaryan can we get this merged! Thanks.

**Comment by tigrannajaryan** [comment]
@bogdandrutu what's the resolution on this? I was not present in the SIG meeting where you discussed this (the meeting notes say that it was discussed, but doesn't tell what was decided).

**Comment by hossain-rayhan** [comment]
@bogdandrutu We need to get this merged today (05/10). Can you please merge this. Thanks.
cc: @alolita

**Comment by hossain-rayhan** [comment]
I just rebased the code so that it passes windows/trace unit tests.

**Pull request merged**